File: analyze_unused_files.php
<?php

$root = '/Users/admin/Documents/GeekBoard';
$excludeDirs = ['.git', 'node_modules', 'vendor', '.gemini', '.vscode', 'fpdf_temp', 'logs'];
$extensions = ['php', 'html', 'js', 'css'];

$files = [];
$references = [];

// 1. Scan Directory
$iterator = new RecursiveIteratorIterator(
    new RecursiveDirectoryIterator($root, RecursiveDirectoryIterator::SKIP_DOTS)
);

foreach ($iterator as $file) {
    if ($file->isFile()) {
        $path = $file->getPathname();
        $relativePath = str_replace($root . '/', '', $path);
        
        // Check excludes
        $exclude = false;
        foreach ($excludeDirs as $dir) {
            if (strpos($relativePath, $dir . '/') === 0) {
                $exclude = true;
                break;
            }
        }
        if ($exclude) continue;

        $ext = $file->getExtension();
        $files[$relativePath] = [
            'path' => $relativePath,
            'ext' => $ext,
            'size' => $file->getSize(),
            'mtime' => $file->getMTime(),
            'suspicious' => isSuspicious($relativePath)
        ];

        // 2. Parse for references
        if (in_array($ext, $extensions)) {
            $content = file_get_contents($path);
            
            // PHP Includes/Requires
            preg_match_all('/(include|require)(_once)?\s*[\(\'"\s](.+?)[\'"\)]/i', $content, $matches);
            foreach ($matches[3] as $match) {
                $clean = cleanPath($match);
                if ($clean) $references[$clean][] = $relativePath;
            }

            // Links (href, src, action, url)
            preg_match_all('/(href|src|action|url)\s*=\s*["\']([^"\']+)["\']/', $content, $matches);
            foreach ($matches[2] as $match) {
                 $clean = cleanPath($match);
                 if ($clean) $references[$clean][] = $relativePath;
            }
            
            // Fetch / AJAX
            preg_match_all('/(fetch|open)\s*\(\s*["\']([^"\']+)["\']/', $content, $matches);
            foreach ($matches[2] as $match) {
                $clean = cleanPath($match);
                if ($clean) $references[$clean][] = $relativePath;
            }
             // window.location
            preg_match_all('/window\.location\.href\s*=\s*["\']([^"\']+)["\']/', $content, $matches);
            foreach ($matches[1] as $match) {
                $clean = cleanPath($match);
                if ($clean) $references[$clean][] = $relativePath;
            }
        }
    }
}

function isSuspicious($path) {
    $name = basename($path);
    if (preg_match('/(backup|old|copie|copy|test|temp|tmp|draft|\.bak|\.old|\.save|_\d{4,}|^[a-z]\.php$)/i', $name)) {
        return "Naming convention";
    }
    if (preg_match('/(2|3|4|5|fixed|final|working|new|simple|simplified|corrected|restored)\.(php|html|js|css)$/i', $name)) {
         return "Versioned/Duplicate name";
    }
    return false;
}

function cleanPath($path) {
    // Remove query strings
    $path = explode('?', $path)[0];
    // Remove anchors
    $path = explode('#', $path)[0];
    
    // Simple normalization - this is not perfect but covers most cases
    // We want to match filename.ext
    return basename($path);
}

// 3. Analyze Orphans
$orphans = [];
foreach ($files as $path => $info) {
    $basename = basename($path);
    // If not referenced and not an index/common entry point
    $isReferenced = false;
    
    // Check if basename is in references keys
    if (isset($references[$basename])) {
        $isReferenced = true;
    }
    
    // Fallback: check full relative path matches in references
    // (A bit harder with simple cleanPath logic, but trying to capture 'folder/file.php')
    // We'll rely on basename matching for now as it's most common in PHP includes
    
    if (!$isReferenced) {
        $orphans[] = $path;
    }
}

// 4. Output
$output = [
    'total_files' => count($files),
    'suspicious_files' => array_keys(array_filter($files, fn($f) => $f['suspicious'])),
    'potential_orphans' => $orphans
];

echo json_encode($output, JSON_PRETTY_PRINT);

?>

File: analyze_unused_files.py

import os
import re
import json

ROOT = '/Users/admin/Documents/GeekBoard'
EXCLUDE_DIRS = {'.git', 'node_modules', 'vendor', '.gemini', '.vscode', 'fpdf_temp', 'logs', 'backups', 'backup'}
EXTENSIONS = {'.php', '.html', '.js', '.css'}

files_info = {}
references = set()

suspicious_patterns = [
    r'(backup|old|copie|copy|test|temp|tmp|draft|\.bak|\.old|\.save|_\d{4,}|^[a-z]\.php$)',
    r'(2|3|4|5|fixed|final|working|new|simple|simplified|corrected|restored)\.(php|html|js|css)$'
]

def is_suspicious(filename):
    for pattern in suspicious_patterns:
        if re.search(pattern, filename, re.IGNORECASE):
            return True
    return False

def clean_path(path):
    path = path.split('?')[0]
    path = path.split('#')[0]
    return os.path.basename(path)

# 1. Scan Directory
for root, dirs, files in os.walk(ROOT):
    # Modify dirs in-place to skip excluded
    dirs[:] = [d for d in dirs if d not in EXCLUDE_DIRS]
    
    for file in files:
        file_path = os.path.join(root, file)
        rel_path = os.path.relpath(file_path, ROOT)
        ext = os.path.splitext(file)[1].lower()
        
        if file.startswith('.'):
            continue

        files_info[rel_path] = {
            'path': rel_path,
            'suspicious': is_suspicious(file)
        }
        
        # 2. Parse for references
        if ext in EXTENSIONS:
            try:
                with open(file_path, 'r', encoding='utf-8', errors='ignore') as f:
                    content = f.read()
                    
                    # PHP Includes/Requires
                    # match include 'foo.php', include("foo.php")
                    matches = re.findall(r'(?:include|require)(?:_once)?\s*[\(\'"\s](.+?)[\'"\)]', content, re.IGNORECASE)
                    for m in matches:
                        references.add(clean_path(m))
                        
                    # Links (href, src, action, url)
                    matches = re.findall(r'(?:href|src|action|url)\s*=\s*["\']([^"\']+)["\']', content, re.IGNORECASE)
                    for m in matches:
                        references.add(clean_path(m))
                        
                    # Fetch / AJAX / Window.location
                    matches = re.findall(r'(?:fetch|open|window\.location\.href)\s*[\(=]\s*["\']([^"\']+)["\']', content, re.IGNORECASE)
                    for m in matches:
                        references.add(clean_path(m))
                        
            except Exception as e:
                pass

# 3. Analyze Orphans
orphans = []
for rel_path, info in files_info.items():
    basename = os.path.basename(rel_path)
    # Check if basename is in references
    # Note: this is a weak check (e.g. index.php is usually an entry point, not referenced)
    # But for "useless files" scanning, high chance an unreferenced non-index file is suspicious.
    
    if basename not in references:
        # Filter out common entry points roughly
        if basename in ['index.php', 'login.php', 'logout.php', 'cron.php', 'app.php']:
            continue
        orphans.append(rel_path)

output = {
    'total_files': len(files_info),
    'suspicious_files': [f for f, i in files_info.items() if i['suspicious']],
    'potential_orphans': orphans
}
print(json.dumps(output, indent=2))
